core/domain/user_services.py

```diff
@@ -977,6 +942,8 @@ def update_user_role(user_id, role):
         raise Exception('Role %s does not exist.' % role)
     user_settings = get_user_settings(user_id, strict=True)
     user_settings.role = role
+    if role == feconf.ROLE_ID_GUEST:
```

Comment:
Woah, what is going on here? This seems like a change that should not be happening, we should not be changing a username (a very important identifier!) while we're supposed to be updating a role.

Instead, it should not be possible to update an existing user's role to "guest". Is that actually possible from the /admin Roles page? If so we should fix it.

(Or maybe there is some user flow where this is important that I'm missing ...? I might be overlooking something.)

Reply:
I think guest means a logged out user which means it won't have a username.I checked out the admin page and we cannot update the role to guest. I wanted to test [this line](https://codecov.io/gh/oppia/oppia/src/develop/core/domain/user_services.py#L1012) which depends on username being None. Shouldn't that mean that the user is a guest? If so, that's why I made the change here. If I don't make the change here do you know how can I create a user_id with no username? Thanks!

Reply:
I don't think "username = None" and "user = guest" are correlated. The roles stuff was added a long time after the username stuff.

I think user_id existing and username = None can exist temporarily during the signup process, but you'll need to read through the code to confirm that. Try to figure out when exactly a user_id gets assigned, and when exactly a username gets assigned -- you will need to analyze the code properly here. The main thing I'm fairly sure about is that this has nothing to do with roles.

Reply:
Resolved in the next comment